### ISSUE TITLE
[REVIEW]: h5fortran: object-oriented polymorphic Fortran interface for HDF5 file IO

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2842]
**Submitting author:** @scivision (<a href="http://orcid.org/0000-0002-1637-6526">Michael Hirsch</a>)
**Repository:** <a href="https://github.com/geospace-code/h5fortran" target ="_blank">https://github.com/geospace-code/h5fortran</a>
**Version:** v4.0.0
**Editor:** @pdebuyl
**Reviewer:** @milancurcic, @bast
**Archive:** <a href="https://doi.org/10.5281/zenodo.4946859" target="_blank">10.5281/zenodo.4946859</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/54abd461016730616cbb49beb53436f8/status.svg)](https://joss.theoj.org/papers/54abd461016730616cbb49beb53436f8)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/54abd461016730616cbb49beb53436f8"><img src="https://joss.theoj.org/papers/54abd461016730616cbb49beb53436f8/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/54abd461016730616cbb49beb53436f8/status.svg)](https://joss.theoj.org/papers/54abd461016730616cbb49beb53436f8)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/geospace-code/h5fortran" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@milancurcic & @bast, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @pdebuyl know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/geospace-code/h5fortran">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@scivision) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/geospace-code/h5fortran">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@scivision) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @milancurcic, @bast it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.3733569 is OK
- 10.5281/zenodo.3718418 is OK
- 10.5281/zenodo.3731716 is OK
- 10.1029/2018GL081569 is OK
- 10.1145/1966895.1966900 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: :point_left:

**Comment by pdebuyl** [comment]
@milancurcic @bast make sure to accept the invitation to the reviewers group and to have a look at the reviewer guidelines linked to at the top of this review page.

The review process will happen in this issue page, so questions to the author or to me can be added as comments here.

**Comment by whedon** [comment]
:wave: @milancurcic, please update us on how your review is going.

**Comment by whedon** [comment]
:wave: @bast, please update us on how your review is going.

**Comment by pdebuyl** [comment]
Apparently, our editorial bot is pinging reviews after a week of inactivity. I usually do that "manually", so here goes for the weekly reminder.

**Comment by bast** [comment]
dear Pierre, sorry for silence. I will work on this upcoming weekend. The last few days have been very packed unfortunately.

**Comment by pdebuyl** [comment]
@bast gentle reminder (I act before @whedon comes back!)

**Comment by bast** [comment]
> @bast gentle reminder (I act before @whedon comes back!)

Sorry for being slow. Working on it now and will submit the review before the weekend is over.

**Comment by pdebuyl** [comment]
Thank you @bast for the review!

@scivision there are a number of joss review related issues opened at your repository. With JOSS, the review can happen in a mixed way between the review thread here and issues at the software's repository. You can answer directly in those issues.

In general, I prefer that an explicit notification be sent here for all steps of the review though, as I am not subscribed to the software repository and thus receive no email notification for such events.

**Comment by pdebuyl** [comment]
Hi @scivision , is there any progress on your side?

**Comment by pdebuyl** [comment]
@scivision gentle reminder

**Comment by pdebuyl** [comment]
Hi @scivision do you need a delay with the submission? If yes, it could be paused.

**Comment by scivision** [comment]
Thanks, I responded to each issue/comment on h5fortran repo

**Comment by pdebuyl** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: :point_left:

**Comment by pdebuyl** [comment]
Thank you for the update @scivision 

@milancurcic @bast sorry for the delay here. Could you check the modifications by @scivision and proceed with the review?

**Comment by pdebuyl** [comment]
@milancurcic are you interested in pursuing this review?

**Comment by milancurcic** [comment]
Yes, sorry, I forgot about this thread completely. I'm now back in the loop and will respond shortly.

**Comment by bast** [comment]
Same here, sorry for delay. Picking this up now.

**Comment by milancurcic** [comment]
I think this paper is good to go, thank you all!

**Comment by pdebuyl** [comment]
Thanks @milancurcic for your review!

**Comment by bast** [comment]
I thank the editor and the author for their patience. The paper is good to go from my perspective. Thanks to everybody for a good learning experience also for me.

**Comment by pdebuyl** [comment]
Thanks @bast for your review! Can you go over the checklist? It is part of the review and I can't proceed until all is checked :-)

@scivision I will start to process the "editorial steps" https://joss.readthedocs.io/en/latest/editing.html#after-reviewers-recommend-acceptance already.

**Comment by bast** [comment]
Sorry, I forgot the checklist on top. Now it's done.

**Comment by pdebuyl** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: :point_left:

**Comment by kthyng** [comment]
@pdebuyl Looks like this is almost finished! What is the status here?

**Comment by pdebuyl** [comment]
@kthyng it is now up to me to complete the editorial work. I'll be on it next week or in a week in the worst case.

**Comment by pdebuyl** [comment]
Hi @scivision 

The documentation at https://geospace-code.github.io/h5fortran/ lacks a "narrative documentation", what is typically called the user guide. The README of the project contains most of what's necessary though, could you integrate the README to the FORD documentation output, with install.md and examples.md ? This way, the rendered docs would be complete.

Regarding installation, there is a bit of duplication between the README and the install page. This is fine, but if you look only at install.md you don't see that h5fortran can install an extra instance of HDF5, which can be annoying for many users (HPC users, for instance).

In the README, you mention the lack of support for complex data. Could you mention logical as well? And include both in the article?

Do you think that a mention of [h5xx](https://github.com/fhoefling/h5xx) would be relevant? I don't use it myself but it seems similar in scope: offering a simplified HDF5 API for typical objects (h5xx is based on templates and inspired by h5py).

If there are other research activities that rely on h5fortran, you are welcome to mention them as well.

Finally, I don't see the use case of appending to extensible datasets mentioned. Is it supported by h5fortran? Are there specific routines to ease that process for users? If not, it should be mentioned as a limitation as it is very typical for simulation work.

Sorry for the delay here, Pierre.

**Comment by pdebuyl** [comment]
Hi @scivision could you take the remarks above in consideration?

**Comment by danielskatz** [comment]
👋 @scivision - Any update?  Is this ready to move to being accepted and published?

**Comment by pdebuyl** [comment]
Hi @danielskatz I also notified @scivision by email. The reviewers recommended publication but there are a few pending remarks from my side for which I would like a response.

**Comment by scivision** [comment]
The autobuilding of HDF5 and ZLIB is useful:

* even if the system has working HDF5, having the binaries optimized for your CPU and using the latest version of HDF5 and ZLIB can make HDF5 operations up to multiple times faster
* typically HPC have broken or ABI-incompatible HDF5 libraries

However as in f95daa5f there is a `cmake -Dautobuild=off` option to have CMake configure fail if a compatible HDF5 or ZLIB is not found

**Comment by scivision** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #2842 with the following error: 
```
 Error reading bibliography file paper.bib:
(line 64, column 1):
unexpected "y"
expecting space, ",", white space or "}"
Looks like we failed to compile the PDF

```

**Comment by scivision** [comment]
I  believe I addressed all https://github.com/openjournals/joss-reviews/issues/2842#issuecomment-840427552 in commits just now. Thank you!

**Comment by scivision** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: :point_left:

**Comment by pdebuyl** [comment]
@scivision thank you for the updates

Can you check the version number that is listed in the first comment of this issue page? If it is not up to date, please state what version should be used and provide the zenodo DOI for the archived version of the code.

The metadata of the zenodo archive should have the title "h5fortran: object-oriented polymorphic Fortran interface for HDF5 file IO", only you as "author" and possible other contributors as "contributors" to the entry.

**Comment by pdebuyl** [comment]
PS: I also used the joined citation syntax and filed a PR about it.

**Comment by scivision** [comment]
version 4.0.0
doi: 10.5281/zenodo.4946859
https://zenodo.org/record/4946859

**Comment by pdebuyl** [comment]
@whedon set v4.0.0 as version

**Comment by whedon** [comment]
OK. v4.0.0 is the version.

**Comment by pdebuyl** [comment]
@whedon set 10.5281/zenodo.4946859 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4946859" target="_blank">10.5281/zenodo.4946859</a> is the archive.

**Comment by pdebuyl** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2407

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2407, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by pdebuyl** [comment]
Thank you @scivision for the submission, @milancurcic and @bast for reviewing! The "associate editor-in-chief" in rotation will take over the paper for publication.

Regards,

Pierre

**Comment by bast** [comment]
Thanks a lot also for this very good experience.

**Comment by kyleniemeyer** [comment]
Hi @scivision, I am doing some final checks before publishing. I noticed two minor issues, could you please resolve theses?

- [ ] In two places (the final paragraph of Summary, and at the end of the Statement of Need) you have references that should use the `[@author1:2001; @author2:2001]` citation syntax, rather than ``[@author1:2001][@author2:2001]` as you have now (this will put them in a single citation separated by commas)
- [ ] In the Collette 2013 reference, please remove the quotation marks around O'Reilly Media, Inc. You might also add the ISBN via the field `isbn = {9781449367831},`

**Comment by scivision** [comment]
Thanks I pushed those changes

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02842/joss.02842/10.21105.joss.02842.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1408568545089363971) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2420
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.02842](https://doi.org/10.21105/joss.02842)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congratulations @scivision on your article's publication in JOSS!

Many thanks to @milancurcic and @bast for reviewing this, and @pdebuyl for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02842/status.svg)](https://doi.org/10.21105/joss.02842)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02842">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02842/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02842/status.svg
   :target: https://doi.org/10.21105/joss.02842
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02842/status.svg)](https://doi.org/10.21105/joss.02842)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
